File: TheaterSuggestions/CSharp/SeatsSuggestions/DeepModel/OfferingAdjacentSeatsToMembersOfTheSameParty.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace SeatsSuggestions.DeepModel
{
    /// <summary>
    ///     Business Rule: offer adjacent seats to members of the same party
    /// </summary>
    public static class OfferingAdjacentSeatsToMembersOfTheSameParty
    {
        private static AdjacentSeats NoAdjacentSeatFound { get; } = new(new List<SeatWithTheDistanceFromTheMiddleOfTheRow>());

        public static IEnumerable<Seat>
            OfferAdjacentSeats(
                SuggestionRequest suggestionRequest,
                IEnumerable<SeatWithTheDistanceFromTheMiddleOfTheRow> seatsWithDistanceFromTheMiddleOfRow)
        {
            var splitInGroupsOfAdjacentSeats = SplitInGroupsOfAdjacentSeats(seatsWithDistanceFromTheMiddleOfRow);

            var offerAdjacentSeats = SelectAdjacentSeatsWithShorterDistanceFromTheMiddleOfTheRow(
                    suggestionRequest,
                    splitInGroupsOfAdjacentSeats)
                .SeatsWithDistance
                .Select(s => s.Seat);

            return offerAdjacentSeats;
        }

        private static AdjacentSeats
            SelectAdjacentSeatsWithShorterDistanceFromTheMiddleOfTheRow(
                SuggestionRequest suggestionRequest,
                AdjacentSeatsGroups groupOfAdjacentSeats)
        {
            var theBestDistancesNearerTheMiddleOfTheRowPerGroup =
                new SortedDictionary<int, AdjacentSeatsGroups>();

            // To select the best group of adjacent seats, we sort them by their distances
            foreach (var adjacentSeats in SortGroupsByDistanceFromMiddleOfTheRow(suggestionRequest,
                groupOfAdjacentSeats).Groups)
            {
                if (!IsMatchingPartyRequested(suggestionRequest, adjacentSeats.SeatsWithDistance)) continue;

                var sumOfDistances = adjacentSeats.SeatsWithDistance.Sum(s => s.DistanceFromTheMiddleOfTheRow);

                if (!theBestDistancesNearerTheMiddleOfTheRowPerGroup.ContainsKey(sumOfDistances))
                {
                    var adjacentSeatsGroups = new AdjacentSeatsGroups();
                    theBestDistancesNearerTheMiddleOfTheRowPerGroup[sumOfDistances] = adjacentSeatsGroups;
                }

                theBestDistancesNearerTheMiddleOfTheRowPerGroup[sumOfDistances].Groups.Add(adjacentSeats);
            }

            return theBestDistancesNearerTheMiddleOfTheRowPerGroup.Any()
                ? SelectTheBestGroup(theBestDistancesNearerTheMiddleOfTheRowPerGroup)
                : NoAdjacentSeatFound;
        }

        private static AdjacentSeatsGroups
            SortGroupsByDistanceFromMiddleOfTheRow(SuggestionRequest suggestionRequest,
                AdjacentSeatsGroups groupOfAdjacentSeats)
        {
            var adjacentSeatsGroups = new AdjacentSeatsGroups();

            foreach (var seats in groupOfAdjacentSeats.Groups.Select(adjacentSeats =>
                adjacentSeats.SeatsWithDistance.OrderBy(s => s.DistanceFromTheMiddleOfTheRow)))
                adjacentSeatsGroups.Groups.Add(new AdjacentSeats(seats.Take(suggestionRequest.PartyRequested)));

            return adjacentSeatsGroups;
        }

        private static AdjacentSeats
            SelectTheBestGroup(SortedDictionary<int, AdjacentSeatsGroups> bestGroups)
        {
            return HasOnlyOneBestGroup(bestGroups)
                ? ProjectToAdjacentSeats(bestGroups)
                : DecideWhichGroupIsTheBestWhenDistancesAreEqual(bestGroups);
        }

        private static AdjacentSeats
            DecideWhichGroupIsTheBestWhenDistancesAreEqual(SortedDictionary<int, AdjacentSeatsGroups> bestGroups)
        {
            SortedDictionary<int, AdjacentSeats> decideBetweenIdenticalScores = new();

            foreach (var adjacentSeats in bestGroups
                .Values.SelectMany(adjacentSeatsGroups => adjacentSeatsGroups.Groups))
                SelectTheBestScoreBetweenGroups(decideBetweenIdenticalScores, adjacentSeats);

            return decideBetweenIdenticalScores.LastOrDefault().Value;
        }

        private static void SelectTheBestScoreBetweenGroups(
            SortedDictionary<int, AdjacentSeats> decideBetweenIdenticalScores, AdjacentSeats adjacentSeats)
        {
            if (decideBetweenIdenticalScores.ContainsKey(adjacentSeats.SeatsWithDistance.Count))
            {
                var (bestGroupScore, bestGroupScoreForContained) =
                    ExtractScores(adjacentSeats, decideBetweenIdenticalScores);

                if (bestGroupScore < bestGroupScoreForContained)
                    decideBetweenIdenticalScores[adjacentSeats.SeatsWithDistance.Count] = adjacentSeats;
            }
            else
            {
                decideBetweenIdenticalScores[adjacentSeats.SeatsWithDistance.Count] = adjacentSeats;
            }
        }

        private static (long bestGroupScore, long bestGroupScoreForContained)
            ExtractScores(AdjacentSeats adjacentSeats,
                IReadOnlyDictionary<int, AdjacentSeats> decideBetweenIdenticalScores)
        {
            // Groups are equivalents, the domain expert have decided to select the group with the smallest seat numbers
            // =========================================================================================================
            return (bestGroupScore: adjacentSeats.SeatsWithDistance.Sum(s => s.Seat.Number),
                bestGroupScoreForContained: decideBetweenIdenticalScores[adjacentSeats.SeatsWithDistance.Count]
                    .SeatsWithDistance.Sum(s => s.Seat.Number));
        }

        private static AdjacentSeats ProjectToAdjacentSeats(SortedDictionary<int, AdjacentSeatsGroups> bestGroups)
        {
            return FirstValues(bestGroups).Groups[0];
        }

        private static AdjacentSeatsGroups FirstValues(SortedDictionary<int, AdjacentSeatsGroups> bestGroups)
        {
            return bestGroups.Values.First();
        }


        private static bool HasOnlyOneBestGroup(SortedDictionary<int, AdjacentSeatsGroups> bestGroups)
        {
            return bestGroups.FirstOrDefault().Value.Groups.Count == 1;
        }

        private static bool IsMatchingPartyRequested(SuggestionRequest suggestionRequest, ICollection seatWithDistances)
        {
            return seatWithDistances.Count >= suggestionRequest.PartyRequested;
        }

        private static AdjacentSeatsGroups SplitInGroupsOfAdjacentSeats(
            IEnumerable<SeatWithTheDistanceFromTheMiddleOfTheRow> seatsWithDistances)
        {
            var adjacentSeats = new AdjacentSeats(new List<SeatWithTheDistanceFromTheMiddleOfTheRow>());
            var groupsOfAdjacentSeats = new AdjacentSeatsGroups();

            using (var enumerator = seatsWithDistances.OrderBy(s => s.Seat.Number).GetEnumerator())
            {
                SeatWithTheDistanceFromTheMiddleOfTheRow seatWithTheDistancePrevious = null;

                while (enumerator.MoveNext())
                {
                    var seatWithDistance = enumerator.Current;
                    if (seatWithTheDistancePrevious == null)
                    {
                        seatWithTheDistancePrevious = seatWithDistance;
                        adjacentSeats.AddSeat(seatWithTheDistancePrevious);
                    }
                    else
                    {
                        if (seatWithDistance?.Seat.Number == seatWithTheDistancePrevious.Seat.Number + 1)
                        {
                            adjacentSeats.AddSeat(seatWithDistance);
                            seatWithTheDistancePrevious = seatWithDistance;
                        }
                        else
                        {
                            groupsOfAdjacentSeats.Groups.Add(adjacentSeats);
                            adjacentSeats = new AdjacentSeats(new List<SeatWithTheDistanceFromTheMiddleOfTheRow>() {seatWithDistance});
                            seatWithTheDistancePrevious = null;
                        }
                    }
                }
            }

            groupsOfAdjacentSeats.Groups.Add(adjacentSeats);

            return groupsOfAdjacentSeats;
        }
    }
}

File: TheaterSuggestions/Java/SeatsSuggestionsDomain/src/main/java/com/baasie/SeatsSuggestionsDomain/DeepModel/OfferingAdjacentSeatsToMembersOfTheSameParty.java
package com.baasie.SeatsSuggestionsDomain.DeepModel;

import com.baasie.SeatsSuggestionsDomain.Seat;
import com.baasie.SeatsSuggestionsDomain.SuggestionRequest;

import java.util.*;
import java.util.stream.Collectors;

public class OfferingAdjacentSeatsToMembersOfTheSameParty {
    private static final AdjacentSeats noSeatFound = new AdjacentSeats();

    private static AdjacentSeats
    selectTheBestGroup(TreeMap<Integer, AdjacentSeatsGroups> bestGroups) {

        return hasOnlyOneBestGroup(bestGroups) ?
                projectToSeats(bestGroups) :
                decideWhichGroupIsTheBestWhenDistancesAreEqual(bestGroups);
    }

    private static AdjacentSeats
    decideWhichGroupIsTheBestWhenDistancesAreEqual(TreeMap<Integer, AdjacentSeatsGroups> bestGroups) {

        TreeMap<Integer,AdjacentSeats> decideBetweenIdenticalScores = new TreeMap<>();

        for (var adjacentSeatsGroups : bestGroups.values())
            SelectTheBestScoreBetweenGroups(decideBetweenIdenticalScores, adjacentSeatsGroups);

        return decideBetweenIdenticalScores.lastEntry().getValue();
    }

    private static void
    SelectTheBestScoreBetweenGroups(TreeMap<Integer, AdjacentSeats> decideBetweenIdenticalScores,
                                    AdjacentSeatsGroups adjacentSeatsGroups) {

        for (var adjacentSeats : adjacentSeatsGroups.groups) {
            if (!decideBetweenIdenticalScores.containsKey(adjacentSeats.seatsWithDistance.size())) {
                decideBetweenIdenticalScores.put(adjacentSeats.seatsWithDistance.size(), adjacentSeats);
            }
        }
    }

    private static AdjacentSeats
    projectToSeats(TreeMap<Integer, AdjacentSeatsGroups> bestGroups) {
        return bestGroups.get(bestGroups.firstKey()).groups.get(0);
    }

    private static boolean
    hasOnlyOneBestGroup(TreeMap<Integer, AdjacentSeatsGroups> bestGroups) {
        return bestGroups.get(bestGroups.firstKey()).groups.size() == 1;
    }

    private static AdjacentSeatsGroups
    splitInGroupsOfAdjacentSeats(SuggestionRequest suggestionRequest, List<SeatWithDistance> seatsWithDistances) {

        var adjacentSeats = new AdjacentSeats();
        var  adjacentSeatsGroups = new AdjacentSeatsGroups();
        SeatWithDistance seatWithTheDistancePrevious = null;

        for (SeatWithDistance seat : orderSeatsByTheirNumberToGroupAdjacent(seatsWithDistances)) {
            if (seatWithTheDistancePrevious == null) {
                seatWithTheDistancePrevious = seat;
                adjacentSeats.addSeat(seatWithTheDistancePrevious);
            } else {
                if (seat.seat().number() == seatWithTheDistancePrevious.seat().number() + 1) {
                    adjacentSeats.addSeat(seat);
                    seatWithTheDistancePrevious = seat;
                } else {
                    adjacentSeatsGroups.groups.add(buildAdjacentSeatsSortedByDistanceFromMiddleOfTheRow(adjacentSeats));
                    adjacentSeats = new AdjacentSeats();
                    adjacentSeats.addSeat(seat);
                    seatWithTheDistancePrevious = null;
                }
            }
        }
        adjacentSeatsGroups.groups.add(buildAdjacentSeatsSortedByDistanceFromMiddleOfTheRow(adjacentSeats));

        return sortGroupsByDistanceFromMiddleOfTheRow(suggestionRequest, adjacentSeatsGroups);
    }

    private static AdjacentSeatsGroups
    sortGroupsByDistanceFromMiddleOfTheRow(SuggestionRequest suggestionRequest,
                                           AdjacentSeatsGroups groupOfAdjacentSeats)
    {
        var adjacentSeatsGroups = new AdjacentSeatsGroups();

        for (var adjacentSeats : new ArrayList<>(groupOfAdjacentSeats.groups)) {
            List<SeatWithDistance> seatWithDistances = adjacentSeats.seatsWithDistance.stream().limit(suggestionRequest.partyRequested().partySize()).collect(Collectors.toList());
            adjacentSeatsGroups.groups.add(new AdjacentSeats(seatWithDistances.stream().sorted(Comparator.comparing(SeatWithDistance::distanceFromTheMiddleOfTheRow)).collect(Collectors.toList())));
        }
        return adjacentSeatsGroups;
    }
    private static AdjacentSeats
    buildAdjacentSeatsSortedByDistanceFromMiddleOfTheRow(AdjacentSeats adjacentSeats) {
        return new AdjacentSeats(adjacentSeats.seatsWithDistance.stream()
                .sorted(Comparator.comparing(SeatWithDistance::distanceFromTheMiddleOfTheRow)).collect(Collectors.toList()));
    }

    private static boolean
    any(TreeMap<Integer, AdjacentSeatsGroups> bestGroups) {
        return !bestGroups.isEmpty();
    }

    private static boolean
    isMatchingPartyRequested(SuggestionRequest suggestionRequest, AdjacentSeats adjacentSeats) {
        return adjacentSeats.seatsWithDistance.size() >= suggestionRequest.partyRequested().partySize();
    }

    private static List<SeatWithDistance>
    orderSeatsByTheirNumberToGroupAdjacent(List<SeatWithDistance> seatsWithDistances) {
        return seatsWithDistances.stream()
                .sorted(Comparator.comparing(s -> s.seat().number()))
                .collect(Collectors.toList());
    }

    public static List<Seat>
    OfferAdjacentSeats(SuggestionRequest suggestionRequest, List<SeatWithDistance> seatsWithDistances) {

       var groupOfAdjacentSeats = splitInGroupsOfAdjacentSeats(suggestionRequest, seatsWithDistances);

        return adaptAdjacentSeats(selectAdjacentSeatsWithShorterDistanceFromTheMiddleOfTheRow(suggestionRequest, groupOfAdjacentSeats));
    }

    private static List<Seat>
    adaptAdjacentSeats(AdjacentSeats adjacentSeats)
    {
        return adjacentSeats.seatsWithDistance.stream().map(SeatWithDistance::seat).collect(Collectors.toList());
    }

    private static AdjacentSeats
    selectAdjacentSeatsWithShorterDistanceFromTheMiddleOfTheRow(SuggestionRequest suggestionRequest, AdjacentSeatsGroups adjacentSeatsGroups) {

        TreeMap<Integer, AdjacentSeatsGroups> theBestDistancesNearerToTheMiddleOfTheRowPerGroup = new TreeMap<>();

        // To select the best group of adjacent seats, we sort them by their distances
        for (var adjacentSeats : adjacentSeatsGroups.groups) {

            if (!isMatchingPartyRequested(suggestionRequest, adjacentSeats)) continue;

            var sumOfDistances = sumOfDistancesNearerTheMiddleOfTheRowPerSeat(adjacentSeats);

            if (!theBestDistancesNearerToTheMiddleOfTheRowPerGroup.containsKey(sumOfDistances))
                theBestDistancesNearerToTheMiddleOfTheRowPerGroup.put(sumOfDistances, new AdjacentSeatsGroups());

            theBestDistancesNearerToTheMiddleOfTheRowPerGroup.get(sumOfDistances).groups.add(adjacentSeats);
        }

        return any(theBestDistancesNearerToTheMiddleOfTheRowPerGroup)
                ? selectTheBestGroup(theBestDistancesNearerToTheMiddleOfTheRowPerGroup) : noSeatFound;
    }

    private static Integer
    sumOfDistancesNearerTheMiddleOfTheRowPerSeat(AdjacentSeats adjacentSeats) {

        return adjacentSeats.seatsWithDistance.stream()
                .map(SeatWithDistance::distanceFromTheMiddleOfTheRow)
                .reduce(0, Integer::sum);
    }

}
